src/cmdlib.sh

```diff
@@ -127,10 +144,105 @@ EOF
     fi
 
     rm -f ${changed_stamp}
-    set -x
-    sudo rpm-ostree compose tree --repo=${workdir}/repo-build --cachedir=${workdir}/cache \
-         --touch-if-changed "${changed_stamp}" \
-         ${treecompose_args} \
-         ${TREECOMPOSE_FLAGS:-} ${manifest} "$@"
-    set +x
+
+    # really, we don't need to do this whole jazz for `fetch`, but meh, it's
+    # just some minimal I/O in that case
+    cat > ${TMPDIR}/compose.sh <<EOF
+SUDO=
+if [ \$(id -u) != 0 ]; then
+    SUDO=sudo
+fi
+
+# build in a temp bare-user repo in cache/ so we can hardlink from pkgcache
+\$SUDO ostree --repo=${workdir}/cache/repo-build init --mode=bare-user
+
+# seed with commit metadata of last build
+# TODO: don't rely on ${ref}
+if ostree --repo=${workdir}/repo rev-parse ${ref} 2>/dev/null; then
+    \$SUDO ostree --repo=${workdir}/cache/repo-build pull-local \
+        ${workdir}/repo ${ref} --commit-metadata-only
+fi
+
+\$SUDO rpm-ostree compose tree --repo=${workdir}/cache/repo-build \
+    --cachedir=${workdir}/cache --touch-if-changed "${changed_stamp}" \
+    ${treecompose_args} ${TREECOMPOSE_FLAGS:-} \
+    ${manifest} $@
+
+\$SUDO chmod -R a+rX ${workdir}/cache/repo-build/objects
+# can't use ${changed_stamp}; it has different semantics with --download-only
+if ostree --repo=${workdir}/cache/repo-build rev-parse ${ref} 2>/dev/null; then
+    ostree --repo=${workdir}/repo pull-local ${workdir}/cache/repo-build ${ref}
+fi
+
+# just nuke it instead of pruning, it's fast to rebuild from pkgcache hardlinks
+\$SUDO rm -rf ${workdir}/repo-build
```

Comment:
i thought we moved to `${workdir}/cache/repo-build` - also i thought we didn't need `repo-build` at all anymore outside of the cache.qcow?